vermouth/processors/generate_coordinates.py

```diff
@@ -0,0 +1,397 @@
+"""
+
+import numpy as np
+import networkx as nx
+
+from .processor import Processor
+
+
+def mindist(A, B):
+    return ((A[:, None] - B[None])**2).sum(axis=2).min(axis=1)
+
+
+def get_missing_atoms(molecule):
+    '''
+    Determine particles without coordinates in molecule.
+    '''
+    return [
+        ix for ix in molecule
+        if molecule.nodes[ix].get('position') is None
+    ]
+
+
+def get_anchors(molecule, indices):
+    '''
+    Determine particles with known coordinates connected to
+    particles with given indices.
+    '''
+    return {
+        a for ix in indices for a in molecule[ix].keys()
+        if molecule.nodes[a].get('position') is not None
+    }
+
+
+def align_z(v):
+    '''
+    Generate rotation matrix aligning z-axis onto v (and vice-versa).
+    '''
+
+    # This routine is based on the notion that for any two
+    # vectors, the alignment is a 180 degree rotation
+    # around the resultant vector.
+
+    w = v / (v ** 2).sum() ** 0.5
+
+    if w[2] <= 1e-8 - 1:
+        return -np.eye(3) # Mind the inversion ...
+
+    w[2] += 1
+
+    return (2 / (w**2).sum()) * w * w[:, None] - np.eye(3)
+
+
+def conicring(n=24, distance=0.125, angle=109.4, axis=None):
+    '''
+    Create a ring of positions corresponding to a cone with given angle.
+    An angle of 109.4 is suitable for ideal SP3 substituents.
+    '''
+    p = 2 * np.pi * np.arange(n) / n
+    r = distance * np.sin(angle * np.pi / 360)
+    z = distance * np.cos(angle * np.pi / 360)
+    X = np.stack([r * np.cos(p), r * np.sin(p), z * np.ones(n)], axis=1)
+    if axis is None:
+        return X
+    return X @ align_z(axis)
+
+
+def double_cone(n=24, distance=0.125, angle=109.4, axis=None):
+    '''Create positions for two consecutive (SP3) bonded particles'''
+    P = conicring(n, distance, angle, axis)
+    S = np.array([P @ align_z(x) + x for x in P])
+    return P, S
+
+
+def triple_cone(n=24, distance=0.125, angle=109.4):
+    '''Create possible positions for three consecutive (SP3) bonded particles'''
+    P, S = double_cone(n, distance, angle)
+    T = np.array([[[P @ align_z(u - x) + u] for u in U] for x, U in zip(P, S)])
+    return P, S, T
+
+
+def _out(molecule, anchor, target=None, control=None, distance=0.125):
+    '''
+    Generate coordinates for atom bonded to anchor
+    using resultant vector of known substituents.
+
+    This approach works for placement of single unknown
+    substituents on sp1, sp2 or sp3 atoms. It should also
+    work for bipyramidal and other configurations.
+
+     b1
+       \
+    b2--a-->X
+       /
+     b3
+
+    Parameters
+    ----------
+
+    Returns
+    -------
+        None
+    '''
+    a = molecule.nodes[anchor]['position']
+    if control == None:
+        B = [
+            molecule.nodes[ix]['position']
+            for ix, v in molecule[anchor].items()
+            if v != {}
+        ]
+    else:
+        B = [ molecule.nodes[ix]['position'] for ix in control ]
```

Comment:
I suppose that the shape of B determines how the atom is rebuild? 

Now the default is to get B from the anchor, which otherwise you get from indices. For making the code more readable and testable, I think it would be better to either accept anchors or only accept indices. It seems redundant to have both. Maybe I'm mistaken, but I'm happy to learn.

Reply:
Agreed. It's a remnant, but I was afraid of removing it in case it proved useful later.